### PR TITLE
Add CCD API to create case callback

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Create case: add endpoint to orchestrator](https://tools.hmcts.net/jira/browse/BPS-741)

### Change description ###

Core implementation minimal version. As mentioned previously, unit testing was and would be insane if covering everything after including ccd. Therefore basing this on top of endpoint #587 

More to come:

- integration test preferred not unit test
- cleanup
- config (to include transformation urls
- definition files (if covered in the ticket)
- pray for acceptance (actually - praying right now)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 1]
```diff
@@ -39,17 +43,20 @@
...
     private final AuthTokenGenerator s2sTokenGenerator;
+    private final CoreCaseDataApi feignCcdApi;
 
     public CreateCaseCallbackService(
         CreateCaseValidator validator,
         ServiceConfigProvider serviceConfigProvider,
         TransformationClient transformationClient,
-        AuthTokenGenerator s2sTokenGenerator
+        AuthTokenGenerator s2sTokenGenerator,
+        CoreCaseDataApi feignCcdApi
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: How about `ccdApi`?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 2]
```diff
@@ -100,30 +109,50 @@ public CreateCaseCallbackService(
         ExceptionRecord exceptionRecord,
         ServiceConfigItem configItem,
         long caseId,
-        boolean ignoreWarnings
+        boolean ignoreWarnings,
+        String idamToken,
+        String userId
     ) {
+        String caseIdStringify = Long.toString(caseId);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Best to make it a nount, e.g. `caseIdString`, `caseIdAsString` or sth like this.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 3]
```diff
@@ -100,30 +109,50 @@ public CreateCaseCallbackService(
...
+
+        log.info("Start creating exception record for {} {}", configItem.getService(), caseIdStringify);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Worth adding `service` and `case ID` into log message. I can see it was already like this, not sure why you changed it.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java (hunk 1 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Mixes up with our own CcdApi in codebase. This reeks for general review and unification. tl;dr - could not re-use CcdApi. But perhaps for this class purposes - can be renamed, np

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java (hunk 3 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Perhaps because code was done long since, just rebased. Sorry about that. Can add now but will review logging in future PR

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java [hunk 4]
```diff
@@ -27,9 +28,7 @@
 @IntegrationTest
 class CreateCaseCallbackTest {
 
-    private static final String CASE_ID_REGEX =
-        // will be replaced with real one after ccd api
-        "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$";
+    private static final String CASE_ID_REGEX = "\\d+";
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java]: Wy don't we verify that values are exactly the same?

Comment (doncem) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java]: Let me update this then. Was quickly renaming it but intension was to actually copy from json files. I'll do that and add a comment

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 5]
```diff
@@ -100,30 +109,54 @@ public CreateCaseCallbackService(
...
 
             SuccessfulTransformationResponse transformationResponse = transformationClient.transformExceptionRecord(
                 configItem.getTransformationUrl(),
                 exceptionRecord,
-                s2sTokenGenerator.generate()
+                s2sToken
             );
 
             if (!ignoreWarnings && !transformationResponse.warnings.isEmpty()) {
                 log.warn("Transformation warnings: {}", String.join(", ", transformationResponse.warnings));
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: I know it's not this PR, but worth double checking if we can log warnings from transformation service. @nitinprabhu ?

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: actually warnings may include sensitive Info so it would be safe to not log it

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: gone 🐩

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 6]
```diff
@@ -70,7 +77,9 @@ public CreateCaseCallbackService(
                     exceptionRecord,
                     configItem,
                     request.getCaseDetails().getId(),
-                    request.isIgnoreWarnings()
+                    request.isIgnoreWarnings(),
+                    idamToken,
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: hope we are passing this idamToken - one which we recieved from callback ?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java [hunk 7]
```diff
@@ -18,19 +19,15 @@
...
 import static org.hamcrest.Matchers.hasItem;
 import static org.hamcrest.Matchers.hasItems;
 import static org.hamcrest.Matchers.nullValue;
-import static org.hamcrest.text.MatchesPattern.matchesPattern;
 import static org.springframework.http.HttpStatus.OK;
 
 @IntegrationTest
 class CreateCaseCallbackTest {
 
-    private static final String CASE_ID_REGEX =
```
Comment (nitinprabhu) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java]: nice we didn't needed regex in first instance 👍

File: src/integrationTest/resources/ccd/response/start-event.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (nitinprabhu) [inline src/integrationTest/resources/ccd/response/start-event.json]: worth renaming this file to start-event-response but can be done later

Comment (doncem) [reply src/integrationTest/resources/ccd/response/start-event.json]: it's already under `response` folder. no need to be too verbose here with resources

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 9]
```diff
@@ -100,30 +109,53 @@ public CreateCaseCallbackService(
...
                 exceptionRecord,
-                s2sTokenGenerator.generate()
+                s2sToken
             );
 
             if (!ignoreWarnings && !transformationResponse.warnings.isEmpty()) {
-                log.warn("Transformation warnings: {}", String.join(", ", transformationResponse.warnings));
-
+                // do not log warnings
                 return Validation.invalid(Array.ofAll(transformationResponse.warnings));
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Maybe not necessarily for this PR, but this method needs to start returning a different kind of result. Currently, it returns:
* valid when successful response
* valid when errors
* invalid when warnings

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java [hunk 10]
```diff
@@ -40,12 +37,16 @@
...
 
         postWithBody(getRequestBody("valid-new-application-with-ocr.json"))
             .statusCode(OK.value())
             .body("errors", empty())
             .body("warnings", empty())
-            .body("data.caseReference", matchesPattern(CASE_ID_REGEX));
+            .body("data.caseReference", equalTo("1539007368674134")); // from sample-case.json
     }
 
     @Test
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java]: I've just noticed that this test expects the body to contain warnings in `errors` field.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java (hunk 9 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: There's inconsistency I agree, but on the ccd it ends up in a same way - error'ed. There's a tiny possibility to achieve errors + warnings with this new design, but essentially it'll end up as errors disregardingly. The way it was before with another callback endpoint

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java (hunk 10 above)
Comment (doncem) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/controllers/CreateCaseCallbackTest.java]: Let me return a warning then. Makes sense. It all comes down to this "proper way of handling response from service" matter

**Comment by lgonczar** [comment]
There are two, in my opinion, important things to test somewhere:
* have we sent the right data to transformation service, i.e. consistent with the data that we received with the callback?
* have we created the case with data from transformation service?

Do we have the above tested somewhere?

**Comment by doncem** [comment]
This is explicitly spec'ing out hardcoded data aka double hardcoding the verification. If it is a blocker - fair thing, can add. If it is not then all static tests cover edge cases where necessary.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 11]
```diff
@@ -100,30 +110,56 @@ public CreateCaseCallbackService(
...
+                s2sToken
             );
 
             if (!ignoreWarnings && !transformationResponse.warnings.isEmpty()) {
-                log.warn("Transformation warnings: {}", String.join(", ", transformationResponse.warnings));
-
-                return Validation.invalid(Array.ofAll(transformationResponse.warnings));
-            } else {
+                // do not log warnings
                 return Validation.valid(new ProcessResult(
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: I think we don't have any `Validation.invalid` case anymore, so this method can return ProcessResult. Can be done in a future PR.

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Yep. part of whichever cleanup I'm planning to do. Ta

**Comment by doncem** [comment]
last changes only affect static tests. they passed. merging. 🕔
